### ISSUE TITLE
Make testdriver blow up sooner with a nice message when trying to do two calls at the same time.

### ISSUE THREAD

**Issue by kereliuk** [w3c/web-platform-tests#10990]
As far as I know this isn't really enforced in a descriptive way at the moment.

**Comment by jugglinmike** [comment]
I'm only now seeing this issue, though, and the wording makes me think that there is some fundamental reason why we can't support that. Care to fill me in, @kereliuk?

**Comment by jugglinmike** [comment]
Scratch that--I found https://github.com/web-platform-tests/wpt/pull/10882 and the awesome tutorial you wrote! Nothing there indicates that this restriction is fundamental to the framework, though. If that's true, then my patch should just remove the caveat from the tutorial.

**Comment by kereliuk** [comment]
@jugglinmike You're right, #11202 should fix this. Will review it now!

**Comment by jugglinmike** [comment]
@kereliuk I think gh-11202 is an example of functional fixedness. I saw a technical problem and tried to solve it. I wasn't considering the more fundamental question: why would tests try to executing commands in parallel in the first place?

As far as I know, it's only because they are using the parallelized `async_test`. There's nothing about the testharness.js use case that necessitates it (if there were, then the fundamentally serial WebDriver protocol would be a dead-end).

By allowing parallel invocations, we would simplify the change required to use this library (it would be "safe" to use from `async_test`). That's valuable because larger patches are more susceptible to regression and generally take longer to review.

At the same time, we'd also be allowing tests to be written with less precision. testdriver.js could accommodate inconsistency, but it couldn't solve it. If instead, we intentionally limited testdriver.js as you described above, we'd be promoting better test-writing practices: more explicit execution order reduces potential for flakiness. I think that's worth a little extra churn.

So I'm planning on submitting a patch as you originally outlined in this issue's description. I wanted to get your input before doing so because you've invested time in reviewing gh-11202.

**Comment by foolip** [comment]
It looks like https://github.com/web-platform-tests/wpt/pull/11202 ended by pointing back to this issue. @jugglinmike, can you summarize why queuing up the request isn't the right move here?

**Comment by jugglinmike** [comment]
@foolip Our thinking was that rejecting attempts at parallel execution would encourage authors to write serial tests, and that this would reduce potential for flakiness.

To be honest, though, I am having second (actually third) thoughts. It's not that I am warming up to parallel testing; it's that moving away from parallel testing will take a much more concerted effort. Trying to "encourage" serial execution in an important-but-rarely-used utility may be more frustrating than helpful.

I'm also speaking from some experience integrating the current "fail on parallel execution" version. Of [the autoplay-requiring tests that we want to integrate with testharness.js](https://github.com/web-platform-tests/results-collection/pull/552), many use `async_test` liberally. The patches I have started locally repeat the same scheduling logic I proposed in gh-11202. Beyond a general aversion to duplication, I worry that the additional complexity will make test maintainers more reluctant to accept patches which introduce testharness.js.

This repeated reversal is no doubt annoying for the people helping out with this, probably for none more than @gsnedders. I've been awaiting their triumphant return so we can discuss this synchronously and settle the matter once and for all.

**Comment by foolip** [comment]
I would argue that not all uses of parallel `async_test` are prone to being flaky, although I'll concede that in practice it's easier to make mistakes with many things happening at once.

On a "much more concerted effort", I do find it frustrating that `async_test` and `promise_test` have different behavior, and would like the capability to make `async_test` execute in sequence, perhaps even making that the default.

Even then, however, I'd prefer it if we could make testdriver.js just work for parallel tests.

**Comment by jugglinmike** [comment]
> On a "much more concerted effort", I do find it frustrating that `async_test`
> and `promise_test` have different behavior, and would like the capability to
> make `async_test` execute in sequence, perhaps even making that the default.

See also gh-5970 and gh-6002. I don't know if an opt-in has been discussed,
though. On the one hand, that would facilitate a long-term roll out. On the
other hand, the intermediate state might be a little confusing. This is
fundamental behavior, so just seeing `async_test` wouldn't tell you everything
you need to know about the framework's behavior.

> Even then, however, I'd prefer it if we could make testdriver.js just work
> for parallel tests.

Agreed. I'll make your preference known when I next speak with Geoffrey.

**Comment by gsnedders** [comment]
> Agreed. I'll make your preference known when I next speak with Geoffrey.

My concern is mostly the fact that dispatching events through WebDriver is inherently sequential, so we'd have to queue them somewhere. This isn't per se a blocker, but it could end up in rather weird situations where the document repeatedly gets scrolled up and down to run different actions on different elements as it jumps between tests.

**Comment by foolip** [comment]
If the order in which the commands are queued isn't deterministic that could indeed get a bit funny, but a quite typical case will be that a bunch of `async_test` call `test_driver.bless` the first thing they do, and thus are effectively made to run sequentially by having the commands queue up. Seems OK?

**Comment by jugglinmike** [comment]
> a quite typical case will be that a bunch of `async_test` call `test_driver.bless` the first thing the do, and thus are effectively made to run sequentially by having the commands queue up

I think the effect is even less severe for that case. Sub-tests declared via
`async_test` typically start off running in series (since they're commonly
declared sequentially in one or more `<script>` tags). If the very first thing
they do is execute a function which forces serial execution, the overall
schedule hasn't changed. They will go on to make asynchronous calls, and that's
where the parallelism will be apparent. The use of this API won't change that.

Bare in mind that this logic effects *all* uses of testdriver, not just the
`bless` method. The effect on scheduling will be more apparent for
`async_test`s that are using `click` and/or `send_keys` "deeper" in the test
body. I agree with @gsnedders that this isn't a blocker, but I wanted to make
the distinction clear.

**Comment by mdittmer** [comment]
_friendly ping_. Looks like this issue has already gone through multiple candidate fixes that are still open. @jugglinmike and @gsnedders has there been any progress on this since August?

**Comment by jugglinmike** [comment]
There has not. My aforementioned preference for gh-11202 has not changed.

**Comment by mdittmer** [comment]
_friendly ping_. @gsnedders @jugglinmike is there still a lack of consensus between landing https://github.com/web-platform-tests/wpt/pull/11202 and landing https://github.com/web-platform-tests/wpt/issues/10990 ? If so, how do you propose we make progress on the decision?

**Comment by foolip** [comment]
@LukeZielinski do you know the current state of what happens when you try multiple `test_driver.foo()` calls in parallel? Does it depend on the command and details of the implementation, or does it always work/fail?

It seems to me that as long as the WebDriver HTTP requests initiated by ExecutorWebDriver are blocking and no code up the stack to testdriver.js has unreliable message delivery order, then commands should naturally stack up without any special code to ensure it. If that is already the current state then we can close this.

**Comment by gsnedders** [comment]
> It seems to me that as long as the WebDriver HTTP requests initiated by ExecutorWebDriver are blocking and no code up the stack to testdriver.js has unreliable message delivery order, then commands should naturally stack up without any special code to ensure it. If that is already the current state then we can close this.

That assumes everything is using wptrunner and ExecutorWebDriver, which we currently don't guarantee. :)

**Comment by LukeZielinski** [comment]
@foolip I wasn't sure how this worked so I tried it by changing the [infra test for clicking](https://github.com/web-platform-tests/wpt/blob/master/infrastructure/testdriver/click.html) to have multiple async sub-tests that all call testdriver methods. The test times out in that case. I can see the commands running and completing in the webdriver layer, so probably something is failing up in testdriver/wpt. Empirically, when I have many async_tests, only the last one passes, the remainder time out.

**Comment by jugglinmike** [comment]
gh-11202 is a fix for this; reviewing the patch may help folks understand the technical limitation. In sort: the current implementation is only capable of tracking one testdriver.js "command" at a time. As @LukeZielinski's experimentation suggests, if some previous command is pending at the moment another command is sent, the first will be forgotten.

**Comment by foolip** [comment]
Alright, https://github.com/web-platform-tests/wpt/pull/11202 indeed explains the current behavior.

@LukeZielinski sounds like you had to edit an existing test to test this, do you know if there are real tests that do parallel requests? If there aren't, maybe we should just quickfix this by making parallel requests fail, precisely as the issue title suggests. Could be a 3-liner fix?

**Comment by LukeZielinski** [comment]
It appears (as per #17360) that this won't be a trivial fix to just throw an error in a common place.
